### PR TITLE
Implement code diagnostics in Visual Mode

### PR DESCRIPTION
**Pull request by jmcphers**
### Intent

This change adds support for code diagnostics inside code chunks in Visual Mode. 

image

Addresses https://github.com/rstudio/rstudio/issues/9874.

### Approach

Enabling diagnostics is not that difficult; the hard part is making it sufficiently performant. The existing implementation requires that the document be saved, and the linter runs against the saved copy of the document from the source database. That's a little slow when there's an idle timer and Pandoc process in front of the save. So:

- The diagnostics engine has been modified so that it can run on a code fragment in addition to an entire document.
- The `LintManager`'s dependency on `TextEditingTarget` has been broken; it now takes a `LintSource` that can be either a source editor or a visual chunk.
- The visual chunks can get lint from both the parent editor (when diagnostics for the whole document are retrieved) and from their own embedded `LintManager`. 
- Visual chunks are re-linted when their contents change (after an idle period); this re-lint uses only the content of the chunk (delivered from the front end instead of read from the source database) for performance reasons enumerated above.

The other big part of this change is managing the lint symbols in the "gutter" area to the left of the chunk. We have to manage and position these ourselves, unless the line numbers are turned on for chunks (in which case Ace has a gutter area we can use). 

### Automated Tests

See https://github.com/rstudio/rstudio-ide-automation/issues/106.

### QA Notes

- Very broadly, source diagnostics should now work in visual mode very similarly to source mode.
- We are slightly less aggressive about triggering a scan for problems; editing a chunk (making a change to it) should trigger a diagnostics scan if you're not sure whether one has been run.
- A couple of optional diagnostics settings do not work in the visual editor; specifically, the warnings about defined but not used / used but not defined symbols are not shown. 
- Look out for problems drawing the diagnostics warnings on the correct line, and/or showing the gutter warnings in the right place.
- The "show line numbers in the visual editor" setting affects how we draw warnings as noted above; try testing with it off and on. 

### Checklist

- [x] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [x] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [x] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [x] This PR passes all local unit tests

### CODE REVIEW

File: src/cpp/session/modules/SessionDiagnostics.cpp [hunk 1]
```diff
@@ -796,41 +805,50 @@ Error lintRSourceDocument(const json::JsonRpcRequest& request,
...
-   boost::shared_ptr<SourceDocument> pDoc(new SourceDocument());
-   error = get(documentId, pDoc);
-   
-   // don't log on error here (it's possible that we might attempt to lint a
-   // document immediately after a suspend-resume, and so we fail to get the
-   // contents of that document)
-   if (error)
-      return error;
-   
    FilePath origin = module_context::resolveAliasedPath(documentPath);
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionDiagnostics.cpp:808]: Is this check still meaningful if we're linting an arbitrary chunk of content? Or should content always be thought of as a fragment of some already-existing document?

File: src/gwt/src/org/rstudio/studio/client/workbench/views/output/lint/LintManager.java [hunk 2]
```diff
@@ -14,8 +14,10 @@
  */
 package org.rstudio.studio.client.workbench.views.output.lint;
 
+import com.google.gwt.event.shared.HandlerRegistration;
```
Comment (kevinushey) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/output/lint/LintManager.java:17]: nit: this import should be bundled with other `com.google.gwt` imports

**Review by kevinushey — APPROVED** [review]
LGTM!

File: src/cpp/session/modules/SessionDiagnostics.cpp (hunk 1 above)
Comment (jmcphers) [reply src/cpp/session/modules/SessionDiagnostics.cpp:808]: It's helpful to always think of it as a fragment. It doesn't make a big difference here, though you could imagine that we had a linter that warned you if you referenced a path that didn't exist (relative to the document).

File: src/gwt/src/org/rstudio/studio/client/workbench/views/output/lint/LintManager.java (hunk 2 above)
Comment (jmcphers) [reply src/gwt/src/org/rstudio/studio/client/workbench/views/output/lint/LintManager.java]: Weird, I don't know why the editor did that! I'll move it.

**Pull request merged**